File: raspberrypi-vehicle-sim/src/isotp.py
#!/usr/bin/env python3
"""ISO 15765-2 classic-CAN segmentation for the simulated diagnostic ECUs."""
from __future__ import annotations

from dataclasses import dataclass, field
from canio import Frame, send


@dataclass
class Reassembler:
    expected: int = 0
    next_sequence: int = 1
    payload: bytearray = field(default_factory=bytearray)

    def accept(self, frame: Frame) -> tuple[bytes | None, Frame | None]:
        if not frame.data:
            return None, None
        kind = frame.data[0] >> 4
        if kind == 0:
            length = frame.data[0] & 0x0F
            return frame.data[1:1 + length], None
        if kind == 1:
            self.expected = ((frame.data[0] & 0x0F) << 8) | frame.data[1]
            self.payload = bytearray(frame.data[2:])
            self.next_sequence = 1
            return None, Frame(response_to_request(frame.can_id), b"\x30\x00\x00")
        if kind == 2 and (frame.data[0] & 0x0F) == self.next_sequence:
            self.next_sequence = (self.next_sequence + 1) & 0x0F
            self.payload.extend(frame.data[1:])
            if len(self.payload) >= self.expected:
                result = bytes(self.payload[:self.expected])
                self.reset()
                return result, None
        return None, None

    def reset(self) -> None:
        self.expected, self.next_sequence = 0, 1
        self.payload.clear()


def response_to_request(can_id: int) -> int:
    return can_id - 8 if 0x7E8 <= can_id <= 0x7EF else 0x7E0


def transmit(sock, can_id: int, payload: bytes) -> None:
    """Transmit a response. For the lab, consecutive frames use a conservative 5 ms gap."""
    import time
    if len(payload) <= 7:
        send(sock, Frame(can_id, bytes([len(payload)]) + payload))
        return
    send(sock, Frame(can_id, bytes([0x10 | (len(payload) >> 8), len(payload) & 0xFF]) + payload[:6]))
    sequence, offset = 1, 6
    # A standards-complete peer sends FC. The simulator also waits briefly so simple clients work.
    time.sleep(0.02)
    while offset < len(payload):
        chunk = payload[offset:offset + 7]
        send(sock, Frame(can_id, bytes([0x20 | (sequence & 0x0F)]) + chunk))
        sequence, offset = sequence + 1, offset + 7
        time.sleep(0.005)

File: app/src/main/java/com/raziya/diagnostics/can/IsoTp.kt
package com.raziya.diagnostics.can

class IsoTpReassembler {
    private var expected = 0
    private var nextSequence = 1
    private val payload = mutableListOf<Byte>()

    sealed interface Result {
        data class Complete(val payload: ByteArray) : Result
        data class FlowControlRequired(val frame: CanFrame) : Result
        data object Pending : Result
        data class Invalid(val reason: String) : Result
    }

    fun accept(frame: CanFrame): Result {
        if (frame.data.isEmpty()) return Result.Invalid("Empty frame")
        return when ((frame.data[0].toInt() ushr 4) and 0x0F) {
            0 -> {
                val length = frame.data[0].toInt() and 0x0F
                if (length > frame.data.size - 1) Result.Invalid("Truncated single frame")
                else Result.Complete(frame.data.copyOfRange(1, length + 1))
            }
            1 -> {
                if (frame.data.size < 2) return Result.Invalid("Truncated first frame")
                expected = ((frame.data[0].toInt() and 0x0F) shl 8) or (frame.data[1].toInt() and 0xFF)
                payload.clear()
                payload.addAll(frame.data.drop(2))
                nextSequence = 1
                Result.FlowControlRequired(CanFrame.of(responseToRequestId(frame.id), 0x30, 0x00, 0x00))
            }
            2 -> {
                val sequence = frame.data[0].toInt() and 0x0F
                if (sequence != nextSequence) return reset("Expected CF $nextSequence, got $sequence")
                nextSequence = (nextSequence + 1) and 0x0F
                payload.addAll(frame.data.drop(1))
                if (payload.size >= expected) {
                    val result = payload.take(expected).toByteArray()
                    clear()
                    Result.Complete(result)
                } else Result.Pending
            }
            else -> Result.Invalid("Unsupported ISO-TP frame type")
        }
    }

    private fun responseToRequestId(id: Int) = if (id in 0x7E8..0x7EF) id - 8 else 0x7E0
    private fun reset(reason: String): Result.Invalid { clear(); return Result.Invalid(reason) }
    private fun clear() { expected = 0; nextSequence = 1; payload.clear() }
}
